ddtrace/sampler.py

```diff
@@ -306,218 +295,34 @@ def _parse_rules_from_env_variable(self, rules):
-        # DEV: This means rules should be ordered by the user from most specific to least specific
-        has_configured_rate_limit = self.limiter.rate_limit != DEFAULT_SAMPLING_RATE_LIMIT
-
-        rule_matched = False
-        for rule in self.rules:
-            if rule.matches(span):
-                sampler = rule
-                rule_matched = True
-                break
-
-        if not rule_matched:
-            sampled = super(DatadogSampler, self).sample(span)
-            sampler = self  # type: ignore
-        else:
-            sampled = sampler.sample(span)
-
-        if rule_matched or has_configured_rate_limit:
-            self._set_sampler_decision(span, sampler, sampled)
-
-        # Ensure all allowed traces adhere to the global rate limit
-        allowed = True
-        if sampled:
-            allowed = self.limiter.is_allowed(span.start_ns)
-            if not allowed:
-                self._set_priority(span, USER_REJECT)
-        if has_configured_rate_limit:
-            span.set_metric(SAMPLING_LIMIT_DECISION, self.limiter.effective_rate)
-        return not allowed or sampled
-
-
-class SamplingRule(BaseSampler):
-    """
-    Definition of a sampling rule used by :class:`DatadogSampler` for applying a sample rate on a span
-    """
-
-    NO_RULE = object()
-
-    def __init__(
-        self,
-        sample_rate,  # type: float
-        service=NO_RULE,  # type: Any
-        name=NO_RULE,  # type: Any
-    ):
-        # type: (...) -> None
+    def sample(self, span, allow_false=True):
+        # type: (Span, bool) -> bool
         """
-        Configure a new :class:`SamplingRule`
-
-        .. code:: python
-
-            DatadogSampler([
-                # Sample 100% of any trace
-                SamplingRule(sample_rate=1.0),
-
-                # Sample no healthcheck traces
-                SamplingRule(sample_rate=0, name='flask.request'),
-
-                # Sample all services ending in `-db` based on a regular expression
-                SamplingRule(sample_rate=0.5, service=re.compile('-db$')),
-
-                # Sample based on service name using custom function
-                SamplingRule(sample_rate=0.75, service=lambda service: 'my-app' in service),
-            ])
-
-        :param sample_rate: The sample rate to apply to any matching spans
-        :type sample_rate: :obj:`float` greater than or equal to 0.0 and less than or equal to 1.0
-        :param service: Rule to match the `span.service` on, default no rule defined
-        :type service: :obj:`object` to directly compare, :obj:`function` to evaluate, or :class:`re.Pattern` to match
-        :param name: Rule to match the `span.name` on, default no rule defined
-        :type name: :obj:`object` to directly compare, :obj:`function` to evaluate, or :class:`re.Pattern` to match
+        If allow_false is False, this function will return True regardless of the sampling decision
```

Comment:
Wouldn't this only be called in tracer.py, with `allow_false` = `False` leading to `DatadogSampler.sample()` always returning `True`? That behavior doesn't make sense to me. Shouldn't we care about all of the logic we have around sampling and rate limits etc?

Reply:
The return value of `sample()` is really only used to set the `span.sampled` flag, which is only used internally and doesn't have any effect on outgoing span tags. Regardless of the return value, `sample()` still has all of its intended side effects and outgoing spans still look as expected.

Because of how confusing this is, the `span.sampled` flag is a great candidate for deprecation and removal.

Reply:
ah I see, we only use it internally to basically not send a span to the agent. However we want to usually let the agent do all of the evaluations on the tags we set for a span, not keep the span from ever being sent in the first place. `span.sampled` should essentially just be viewed as `span.send_to_agent`.  What a confusing name. Ok got it thanks!